pkg/util/smallset/smallset.go

```diff
@@ -0,0 +1,110 @@
+// Copyright Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package smallset
+
+import (
+	"fmt"
+
+	"golang.org/x/exp/constraints"
+
+	"istio.io/istio/pkg/slices"
+)
+
+// Set is an immutable set optimized for *small number of items*. For general purposes, Sets is likely better
+//
+// *Set construction*: sets is roughly 1kb allocations per 250 items. smallsets is 0.
+// *Contains* sets is O(1). smallsets is O(logn). smallsets is typically faster up to about 5 elements.
+//
+//	At 1000 items, it is roughly 5x slower (30ns vs 5ns).
+type Set[T constraints.Ordered] struct {
+	items []T
+}
+
+// NewPresorted creates a new Set with the given items.
+// If items is not sorted, this gives undefined behavior; use New instead.
+func NewPresorted[T constraints.Ordered](items ...T) Set[T] {
+	return Set[T]{items: items}
+}
+
+// New creates a new Set with the given items.
+func New[T constraints.Ordered](items ...T) Set[T] {
+	slices.Sort(items)
+	return Set[T]{items: items}
+}
+
+// CopyAndInsert builds a *new* with all the current items plus new items
+func (s Set[T]) CopyAndInsert(items ...T) Set[T] {
+	slices.Sort(items)
+	// This is basically the 'merge' part of merge sort.
+	a := s.items
+	b := items
+	nl := make([]T, 0, len(a)+len(b))
+	i, j := 0, 0
+	appendIfUnique := func(t T) []T {
+		l := len(nl)
+		if l == 0 {
+			nl = append(nl, t)
+		} else {
+			last := nl[l-1]
+			if last != t {
+				nl = append(nl, t)
+			}
+		}
+		return nl
+	}
+	for i < len(a) && j < len(b) {
+		if a[i] < b[j] {
+			nl = appendIfUnique(a[i])
+			i++
+		} else {
+			nl = appendIfUnique(b[j])
+			j++
+		}
+	}
+	for ; i < len(a); i++ {
+		nl = appendIfUnique(a[i])
+	}
+	for ; j < len(b); j++ {
+		nl = appendIfUnique(b[j])
+	}
+	return New(nl...)
```

Comment:
Should this be the presorted new function rather than vanilla new?

Reply:
It seems that using `return Set[T]{items: nl}` can save items copying.